webroot/js/websocket.js

```diff
@@ -0,0 +1,137 @@
+ */
+
+import SOCKET_MESSAGE_TYPES from './chat/socketMessageTypes.js';
+
+const LOCAL_TEST = window.location.href.indexOf('localhost:') >= 0;
+const URL_WEBSOCKET = LOCAL_TEST
+  ? 'wss://goth.land/entry'
+  : `${location.protocol === 'https:' ? 'wss' : 'ws'}://${location.host}/entry`;
+
+const TIMER_WEBSOCKET_RECONNECT = 5000; // ms
+
+const CALLBACKS = {
+  RAW_WEBSOCKET_MESSAGE_RECEIVED: 'rawWebsocketMessageReceived',
+  WEBSOCKET_CONNECTED: 'websocketConnected',
+  WEBSOCKET_DISCONNECTED: 'websocketDisconnected',
+}
+
+class Websocket {
+
+  constructor() {
+    this.websocket = null;
+    this.websocketReconnectTimer = null;
+    this.listeners = [];
+
+    this.send = this.send.bind(this);
+
+    const ws = new WebSocket(URL_WEBSOCKET);
+    ws.onopen = this.onOpen.bind(this);
+    ws.onclose = this.onClose.bind(this);
+    ws.onerror = this.onError.bind(this);
+    ws.onmessage = this.onMessage.bind(this);
+
+    this.websocket = ws;
+  }
+
+  // Other components should register for websocket callbacks.
+  addListener(listener) {
+    this.listeners.push(listener);
+  }
+
+  
+  // Interface with other components
+
+  // Outbound: Other components can pass an object to `send`.
+  send(message) {
+    // Sanity check that what we're sending is a valid type.
+    if (!message.type || !SOCKET_MESSAGE_TYPES[message.type]) {
+      console.warn(`Outbound message: Unknown socket message type: "${message.type}" sent.`);
+    }
+    
+    const messageJSON = JSON.stringify(message);
+    this.websocket.send(messageJSON);
+  }
+
+  // Private methods
+
+  // Fire the callbacks of the listeners.
+  notifyListeners(callbackName, message) {
+    this.listeners.forEach(function (listener) {
+      if (doesObjectSupportFunction(listener, callbackName)) {
+        listener[callbackName](message);
+      }
+    });
+  }
+
+  // Internal websocket callbacks
+
+  onOpen(e) {
+    if (this.websocketReconnectTimer) {
+      clearTimeout(this.websocketReconnectTimer);
+    }
+
+    this.notifyListeners(CALLBACKS.WEBSOCKET_CONNECTED);
+  }
+
+  onClose(e) {
+    // connection closed, discard old websocket and create a new one in 5s
+    this.websocket = null;
+    this.notifyListeners(CALLBACKS.WEBSOCKET_DISCONNECTED);
+    this.handleNetworkingError('Websocket closed.');
+    this.websocketReconnectTimer = setTimeout(this.setupWebsocket, TIMER_WEBSOCKET_RECONNECT);
+  }
+
+  // On ws error just close the socket and let it re-connect again for now.
+  onError(e) {
+    this.handleNetworkingError(`Socket error: ${JSON.parse(e)}`);
+    this.websocket.close();
+  }
+
+  /*
+  onMessage is fired when an inbound object comes across the websocket.
+  We pass these messages along to listeners by firing the
+  RAW_WEBSOCKET_MESSAGE_RECEIVED type callback.
+  If the message is of type `PING` we send a `PONG` back and do not
+  pass it along to listeners.
+  */
+  onMessage(e) {
+    const model = JSON.parse(e.data);
```

Comment:
Let's wrap this in a try/catch in case e.data is not parseable.

Reply:
Done!